### ISSUE TITLE
Support for third-party extendr wrappers

### ISSUE THREAD

**Issue by multimeric** [extendr/extendr#440]
It would be nice to provide a way for the community to create creates that integrate extendr with some other crate, without having to actually live in either `extendr` or that other crate.

For example we have `ndarray::Array`, and say we wanted to `impl TryInto<Robj> for Array` in a third crate rather than in extendr core (where this `impl` currently lives), we would get stopped by the foreign crate rule, since neither Robj nor Array is in this third crate. 

But we could perhaps solve this compiler issue using a new macro. For example if we called it `extendr_third_party`, a developer could:
```rust
#[extendr_third_party] 
impl TryInto<Robj> for Array {}
```
This would:
* automatically newtype `Array`
* put the return value of `try_into` into that newtype
* Unwrap the newtype for the `#[extendr]` function that the end user will write

**Comment by JoJoJet** [comment]
[Originally posted on discord](https://discord.com/channels/751419551132155925/751419551132155928/1080213628759851069):

I haven't fully tried this, but it should be possible to bypass the orphan rule by adding a marker generic to `IntoRobj`:

```rust
pub trait IntoRobj<Marker> { ... }

impl<T: Into<Robj>> IntoRobj<()> for T { ... }

// In a downstream crate...
#[doc(hidden)] pub struct MyMarker;
impl IntoRobj<MyMarker> for SomeForeignType {} // this works because `ForeignTrait<LocalType>` counts as a local trait sometimes
```

This wouldn't be totally seamless since you'd need an extra generic parameter in any function that takes an `IntoRobj` parameter -- however the compiler is able to infer the marker type at the callsite, making this mostly invisible to consumers of the API.

```rust
fn do_a_thing<Marker>(robj: impl IntoRobj<Marker>) { ... }

let x: SomeForeignType = ...;
do_a_thing(x); // the marker type gets inferred automatically
```

This pattern is used in Bevy to prevent overlapping impls, but it also seems to work for bypassing the orphan rule.

**Comment by multimeric** [comment]
Some minor issues I note with the generic conversion trait idea:

* The marker might be confusing to users, considering it's basically a hack around the type system. Creating a unit struct is an odd thing to have to do
* If both conversion traits implement the same conversion, then Rust can't determine which trait to use, and you get an error such as `[E0790] Error: cannot call associated function on trait without specifying the corresponding `impl` type`. This is a confusing error that doesn't relate at all to how is thinking about the conversion process.

Another thought I had for solving this problem: **Customizable conversion trait**. We allow the user to do something like `#[extender(conversion=MyFromRobj)]`. This is nice because it gives full power to the user if they want it, and avoids any orphan rule restrictions. The issue arises when they want to re-use conversions from `extendr` or some other crate. If they `impl MyFromRobj for T where T : FromRobj`, then they can't override the implementations. If we had trait specialization or inheritance this would be a lovely solution but we don't.

But in the end, I'm kind of leaning towards the current behaviour, which is to encourage newtypes. The reason why I think this is okay is that extendr only has to automate the Robj <-> Rust Type conversion. Using newtypes is fine at that point. The only point at which using newtypes is ugly is the user-written Rust implementation, but at that point they can just unwrap the newtype and work with it as normal.

**Comment by multimeric** [comment]
I've written this up in a notebook, which shows the exact compiler warnings we get in each case: https://nbviewer.org/github/multimeric/extendr-conversions/blob/b213254651abe46581c9a1d62b11995bf092d765/conversion.ipynb

**Comment by Ilia-Kosenkov** [comment]
Ok I browsed your notebook, very impressive and in-depth. I am not convinced that 'Overriding (3)' is something we should be concerned with. If we define a conversion, it stays. If someone desperately wants custom conversions for well-known types, go extra mile with new-type or maybe parsing `Robj`. To me, GenericTrait looks somewhat Ok, though I am not sure I fully understand it.

**Comment by multimeric** [comment]
Fair enough, if you think that overriding isn't important, and you aren't concerned by the confusing compiler message when you do, and  think that having "marker structs" as part of our API is okay, then I will put you as a vote for the Generic Trait method.

**Comment by multimeric** [comment]
So far I think @CGMossa is a vote for newtype.

**Comment by Ilia-Kosenkov** [comment]
That's the problem: I do not have enough knowledge to make a decision here. What I can say is that overriding should not be allowed, I consider this harmful. There is a trait `FromStr` for string parsing, and I doubt that whatever implementation exists for `i32` by default should be overridden by someone else. This is basically our case.

**Comment by multimeric** [comment]
As an interesting case study, PyO3:
* Uses custom conversion traits: https://pyo3.rs/v0.18.3/conversions/traits, https://docs.rs/pyo3/latest/pyo3/conversion/index.html
* Suggest newtype for third party structs: https://github.com/PyO3/pyo3/issues/287#issuecomment-441554919

**Comment by multimeric** [comment]
Another case study we have discussed before: Serde.

* Uses custom conversion traits: https://serde.rs/custom-serialization.html
* For third party types, they have this awful system where you have to copy the struct definition, which I personally hate: https://serde.rs/remote-derive.html

**Comment by CGMossa** [comment]
Yeah, but serde also has these:
https://serde.rs/container-attrs.html#bound
https://serde.rs/container-attrs.html#try_from

I've not used or understood the `remote-derive`, but it sounds like the custom trait suggestion.

**Comment by multimeric** [comment]
I'm not sure if `bound` is relevant (?), but `try_from` is interesting. The idea of allowing a two-step conversion, from `Robj` to a wrapper struct (newtype in our case), then to the original type. If you use [shrinkwraprs](https://crates.io/crates/shrinkwraprs) on your newtype then I think this would kind of work for free.
